python/ray/autoscaler/_private/gen2/node_provider.py

```diff
@@ -0,0 +1,758 @@
+                    # currently head node always has floating ip
+                    # in case floating ip present we want to add it
+                    node["floating_ips"] = floating_ips
+
+            nodes.append(node)
+
+            for node in nodes:
+                self.cached_nodes[node["id"]] = node
+
+        return [node["id"] for node in nodes]
+
+    @log_in_out
+    def is_running(self, node_id):
+        with self.lock:
+            node = self._get_cached_node(node_id)
+            return node["status"] == "running"
+
+    @log_in_out
+    def is_terminated(self, node_id):
+        with self.lock:
+            try:
+                node = self._get_cached_node(node_id)
+                return node["status"] not in ["running", "starting", "pending"]
+            except Exception:
+                return True
+
+    @log_in_out
+    def node_tags(self, node_id):
+        with self.lock:
+            return self.nodes_tags.get(node_id, {})
+
+    # return external ip for head and private ips for workers
+    def _get_hybrid_ip(self, node_id):
+        node = self._get_cached_node(node_id)
+        node_type = self._get_node_type(node["name"])
+        if node_type == NODE_KIND_HEAD:
+            fip = node.get("floating_ips")
+            if fip:
+                return fip[0]["address"]
+
+            node = self._get_node(node_id)
+            fip = node.get("floating_ips")
+            if fip:
+                return fip[0]["address"]
+        else:
+            return self.internal_ip(node_id)
+
+    @log_in_out
+    def external_ip(self, node_id):
+        with self.lock:
+            if self.provider_config.get("use_hybrid_ips"):
+                return self._get_hybrid_ip(node_id)
+
+            node = self._get_cached_node(node_id)
+
+            fip = node.get("floating_ips")
+            if fip:
+                return fip[0]["address"]
+
+            node = self._get_node(node_id)
+            fip = node.get("floating_ips")
+            if fip:
+                return fip[0]["address"]
+
+    @log_in_out
+    def internal_ip(self, node_id):
+        node = self._get_cached_node(node_id)
+
+        try:
+            primary_ipv4_address = node["network_interfaces"][0].get(
+                "primary_ipv4_address"
+            )
+            if primary_ipv4_address is None:
+                node = self._get_node(node_id)
+        except Exception:
+            node = self._get_node(node_id)
+
+        logger.debug(f"in internal_ip, returning ip for node {node}")
+
+        return node["network_interfaces"][0].get("primary_ipv4_address")
+
+    @log_in_out
+    def set_node_tags(self, node_id, tags):
+        with self.lock:
+            # update inmemory cache
+            if node_id and tags:
+                node_cache = self.nodes_tags.setdefault(node_id, {})
+                node_cache.update(tags)
+
+            # dump inmemory cache to file
+            ray_cache = Path(Path.home(), Path(".ray"))
```

Comment:
do you mean `Path.home() / ".ray"`? Same for others.

also we should not use `.ray` to avoid collision with other clouds/config files. BTW, why we need to write it to files? It seems other cloud provides do not do this to avoid inconsistency issues (e.g., the tags get out of sync because something happens to local files or in the cloud that is not done by ray autoscaler)

Reply:
Thanks! Fixed. Also moved the tags cache to different file.
Writing to file is a temporary solution. We are working on the solution jointly with Gen2 team. Once few things get improved we will align it with other cloud providers implementations